src/pd/Client.cc

```diff
@@ -184,12 +186,16 @@ void Client::initLeader()
 void Client::updateLeader()
 {
     std::unique_lock lk(leader_mutex);
-    for (const auto & url : urls)
+    for (auto it = urls.begin(); it != urls.end();)
     {
+        const auto & url = *it;
         auto resp = getMembers(url);
         if (!resp.has_header() || resp.leader().client_urls_size() == 0)
         {
-            log->warning("failed to get cluster id by :" + url);
+            log->warning("failed to get cluster id by :" + *it);
+            // move the failed url to the end of the list
+            it = urls.erase(it);
+            urls.push_back(url);
             continue;
         }
         updateURLs(resp.members());
```

Comment:
The function `updateURLs` here will replace `urls` by the list of the resp, which may put the failure pd address in front of the list again.

Reply:
fixed